flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/handler/statement/FetchResultsHandler.java

```diff
@@ -0,0 +1,97 @@
+import org.apache.flink.table.gateway.api.SqlGatewayService;
+import org.apache.flink.table.gateway.api.operation.OperationHandle;
+import org.apache.flink.table.gateway.api.results.ResultSet;
+import org.apache.flink.table.gateway.api.session.SessionHandle;
+import org.apache.flink.table.gateway.api.utils.SqlGatewayException;
+import org.apache.flink.table.gateway.rest.handler.AbstractSqlGatewayRestHandler;
+import org.apache.flink.table.gateway.rest.header.statement.FetchResultsHeaders;
+import org.apache.flink.table.gateway.rest.message.operation.OperationHandleIdPathParameter;
+import org.apache.flink.table.gateway.rest.message.session.SessionHandleIdPathParameter;
+import org.apache.flink.table.gateway.rest.message.statement.FetchResultsResponseBody;
+import org.apache.flink.table.gateway.rest.message.statement.FetchResultsTokenParameters;
+import org.apache.flink.table.gateway.rest.message.statement.FetchResultsTokenPathParameter;
+import org.apache.flink.table.gateway.rest.util.SqlGatewayRestAPIVersion;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+/** Handler to fetch results. */
+public class FetchResultsHandler
+        extends AbstractSqlGatewayRestHandler<
+                EmptyRequestBody, FetchResultsResponseBody, FetchResultsTokenParameters> {
+
+    public FetchResultsHandler(
+            SqlGatewayService service,
+            Map<String, String> responseHeaders,
+            MessageHeaders<EmptyRequestBody, FetchResultsResponseBody, FetchResultsTokenParameters>
+                    messageHeaders) {
+        super(service, responseHeaders, messageHeaders);
+    }
+
+    @Override
+    protected CompletableFuture<FetchResultsResponseBody> handleRequest(
+            SqlGatewayRestAPIVersion version, @Nonnull HandlerRequest<EmptyRequestBody> request) {
+        // Parse the parameters
+        SessionHandle sessionHandle = request.getPathParameter(SessionHandleIdPathParameter.class);
+        OperationHandle operationHandle =
+                request.getPathParameter(OperationHandleIdPathParameter.class);
+        Long token = request.getPathParameter(FetchResultsTokenPathParameter.class);
+
+        // Get the statement results
+        @Nullable ResultSet resultSet;
+        @Nullable String resultType;
+        Long nextToken;
+
+        try {
+            resultSet =
+                    service.fetchResults(sessionHandle, operationHandle, token, Integer.MAX_VALUE);
+            nextToken = resultSet.getNextToken();
+            resultType = resultSet.getResultType().toString();
+        } catch (Exception e) {
+            Throwable root = e;
+            while (root.getCause() != null
+                    && root.getCause().getMessage() != null
+                    && !root.getCause().getMessage().isEmpty()) {
+                root = root.getCause();
+            }
+            throw new SqlGatewayException(root);
```

Comment:
I think we should preserve the complete error stack.
```suggestion
            throw new SqlGatewayException(e);
```